core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java

```diff
@@ -142,4 +115,75 @@ private void doExecute(BlockingRunnable runnable) {
          blockedTasks.offer(runnable);
       }
    }
+
+   private class ControllerThread extends Thread {
+
+      private final LinkedList<BlockingRunnable> readyList;
+      private final Semaphore semaphore;
+      private volatile boolean interrupted;
+
+      public ControllerThread(String controllerThreadName) {
+         super(controllerThreadName);
+         readyList = new LinkedList<>();
+         semaphore = new Semaphore(0);
+      }
+
+      public void checkForReadyTask() {
+         semaphore.release();
+      }
+
+      @Override
+      public void interrupt() {
+         interrupted = true;
+         super.interrupt();
+         semaphore.release();
+
+      }
+
+      @Override
+      public void run() {
+         while (!interrupted) {
+            try {
+               semaphore.acquire();
+            } catch (InterruptedException e) {
+               return;
+            }
+            semaphore.drainPermits();
+            for (Iterator<BlockingRunnable> iterator = blockedTasks.iterator(); iterator.hasNext(); ) {
+               BlockingRunnable runnable = iterator.next();
+               if (runnable.isReady()) {
+                  iterator.remove();
+                  readyList.add(runnable);
+               }
+            }
+
+            if (log.isTraceEnabled()) {
+               log.tracef("Tasks to be executed=%s, still pending=%s", readyList.size(), blockedTasks.size());
+            }
+
+            while (!readyList.isEmpty()) {
```

Comment:
I would recommend to use either an iterator calling remove after each or calling pollFirst in a loop would be much better (latter seems better to me).  Although it won't make that much of a difference if you leave it as is either.